drivers/led_strip/ws2812_i2s.c

```diff
@@ -0,0 +1,255 @@
+    }
+
+    ret = i2s_trigger(cfg->dev, I2S_DIR_TX, I2S_TRIGGER_DRAIN);
+    if (ret < 0) {
+        LOG_ERR("Failed to trigger command %d on TX: %d", I2S_TRIGGER_DRAIN, ret);
+        return ret;
+    }
+
+    /* Wait until transaction is over */
+    flush_time_us = cfg->tx_buf_bytes / (sizeof(uint32_t) * 8);
+    flush_time_us *= 10;        /* 1 LRCK period = 1 32b word */
+    k_usleep(flush_time_us);
+
+    return ret;
+}
+
+static int ws2812_strip_update_channels(const struct device *dev,
+                                        uint8_t *channels,
+                                        size_t num_channels)
+{
+    LOG_ERR("update_channels not implemented");
+    return -ENOTSUP;
+}
+
+static int ws2812_i2s_init(const struct device *dev)
+{
+    const struct ws2812_i2s_cfg *cfg = dev->config;
+    uint8_t i;
+    int ret;
+
+    struct i2s_config config;
+
+    /* 16-bit stereo, 100kHz LCLK */
+    config.word_size = 16;
+    config.channels = 2;
+    config.format = I2S_FMT_DATA_FORMAT_I2S;
+    config.options = I2S_OPT_BIT_CLK_MASTER | I2S_OPT_FRAME_CLK_MASTER;
+    config.frame_clk_freq = WS2812_I2S_LRCK_HZ; /* WS (or LRCK) */
+    config.mem_slab = cfg->mem_slab;
+    config.block_size = cfg->tx_buf_bytes;
+    config.timeout = 1000;
+
+    ret = i2s_configure(cfg->dev, I2S_DIR_TX, &config);
+    if (ret < 0) {
+        LOG_ERR("Failed to configure I2S device: %d\n", ret);
+        return ret;
+    }
+
+    for (i = 0; i < cfg->num_colors; i++) {
+        switch (cfg->color_mapping[i]) {
+            case LED_COLOR_ID_WHITE:
+            case LED_COLOR_ID_RED:
+            case LED_COLOR_ID_GREEN:
+            case LED_COLOR_ID_BLUE:
+                break;
+            default:
+                LOG_ERR("%s: invalid channel to color mapping."
+                        "Check the color-mapping DT property",
+                        dev->name);
+                return -EINVAL;
+        }
+    }
+
+    return 0;
+}
+
+static const struct led_strip_driver_api ws2812_i2s_api = {
+.update_rgb = ws2812_strip_update_rgb,
+.update_channels = ws2812_strip_update_channels,
+};
+
+#define WS2812_RESET_DELAY_US(idx) DT_INST_PROP(idx, reset_delay)
+/* Rounds up to the next 20us. */
+#define WS2812_RESET_DELAY_WORDS(idx) \
+    ((WS2812_RESET_DELAY_US(idx) + (20 - 1)) / 20)
+
+#define WS2812_NUM_COLORS(idx) (DT_INST_PROP_LEN(idx, color_mapping))
+
+#define WS2812_I2S_NUM_PIXELS(idx)              \
+    (DT_INST_PROP(idx, chain_length))
+
+#define WS2812_I2S_BUFSIZE(idx)                                     \
+    ((WS2812_NUM_COLORS(idx) * WS2812_I2S_NUM_PIXELS(idx)           \
+      + WS2812_I2S_PRE_DELAY_WORDS                                  \
+      + WS2812_RESET_DELAY_WORDS(idx))                              \
+     * 4)
+
+#define WS2812_I2S_DEVICE(idx)                                          \
+                                                                        \
+    K_MEM_SLAB_DEFINE_STATIC(ws2812_i2s_##idx##_slab,                   \
```

Comment:
Why bother with this instead of just allocating an array per device?

Reply:
I think it's a requirement of the I2S API (which is not super-flexible)
https://github.com/jori-nordic/zephyr/blob/969d3ffe92e806ffb32fa6d81ecca3cd1f398e09/include/zephyr/drivers/i2s.h#L279

Reply:
another issue with that API is that I can't really check if a transaction's status, hence the sleep duration in the led update implementation.
I could make a retry loop based on the error message of `i2s_write` but I think waiting is cleaner.

If you can stress-test it a bit it would be nice, ie just call the led update function in a loop from your app and see what happens.